### PR TITLE
Flow RazorLanguageVersion to RazorEngine.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Restructured RazorLanguageVersion to be a sealed concrete type to enable things like `RazorLanguageVersion.Latest`; it also allows us to make broader changes in the future. Also, in the future if we want to add support for overriding operators to enable greater than comparisons we can as well.
- Removed version validity checks because we restrict who can construct a `RazorLanguageVersion` now. This way we don't have to check for valid versions all throughout our code.
- Added a simple `ProjectExtensibilityConfiguration` => `RazorLanguageVersion` method in the `DefaultProjectExtensibilityConfigurationFactory` to temporarily enable letting the system operate on the `RazorLanguageVersion`. Eventually that entire class will change.
- Also validated that in VS we now properly detect the RazorLanguageVersion and show appropriate IntelliSense.

#1961 

/cc @mkArtakMSFT

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Language/RazorEngine.cs [hunk 1]
```diff
@@ -15,26 +15,39 @@ public static RazorEngine Create()
             return Create(configure: null);
         }
 
-        public static RazorEngine Create(Action<IRazorEngineBuilder> configure)
+        public static RazorEngine Create(Action<IRazorEngineBuilder> configure) => Create(RazorLanguageVersion.Latest, configure);
+
+        public static RazorEngine Create(RazorLanguageVersion version, Action<IRazorEngineBuilder> configure)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/RazorEngine.cs]: Thoughts on these being `public`? At first I had them `internal` but then I figured that providing the `RazorLanguageVersion` is pretty essential to building a `RazorEngine`. Given that we'll have a similar API for the `RazorProjectEngine` (but will exist on the builder) I wanted to keep a similar API available even if you can't detect it directly via the builder.

File: src/Microsoft.AspNetCore.Razor.Language/RazorLanguageVersion.cs [hunk 2]
```diff
@@ -1,33 +1,33 @@
...
+            Major = major;
+            Minor = minor;
         }
+
+        public int Major { get; }
+
+        public int Minor { get; }
+
+        public override string ToString() => $"Razor Language '{Major}.{Minor}'";
     }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/RazorLanguageVersion.cs]: My vision for this is that when we eventually flow the MSBuild Razor Language version we'll add a method on here to determine or create a proper `RazorLanguageVersion`. Either way I wanted to avoid us having a potentially invalid `RazorLanguageVersion` in our system and therefore having to check `IsValid` in several locations (that's why this is the only class that can construct this class).

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectExtensibilityConfigurationFactory.cs [hunk 3]
```diff
@@ -5,6 +5,7 @@
 using System.Collections.Generic;
```
Comment (NTaylorMullen) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectExtensibilityConfigurationFactory.cs]: This entire class will eventually be re-constructed to support input from MSBuild and build the project extensibility pieces accordingly. For now the changes here are to just get the flow going and setup for success when we move to MSBuild.

File: src/Microsoft.VisualStudio.Editor.Razor/DefaultTemplateEngineFactoryService.cs [hunk 4]
```diff
@@ -15,6 +15,7 @@ namespace Microsoft.VisualStudio.Editor.Razor
     internal class DefaultTemplateEngineFactoryService : RazorTemplateEngineFactoryService
     {
         private readonly static MvcExtensibilityConfiguration DefaultConfiguration = new MvcExtensibilityConfiguration(
+            RazorLanguageVersion.Version2_0,
```
Comment (NTaylorMullen) [inline src/Microsoft.VisualStudio.Editor.Razor/DefaultTemplateEngineFactoryService.cs]: Going to do another pass in a different PR on all of the versions in this class and in `DefaultProjectExtensibilityConfigurationFactory` to update them to 2.1 (latest) since 15.7 aligns with 2.1

File: src/Microsoft.AspNetCore.Razor.Language/RazorEngine.cs [hunk 5]
```diff
@@ -15,26 +15,39 @@ public static RazorEngine Create()
             return Create(configure: null);
         }
 
-        public static RazorEngine Create(Action<IRazorEngineBuilder> configure)
+        public static RazorEngine Create(Action<IRazorEngineBuilder> configure) => Create(RazorLanguageVersion.Latest, configure);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorEngine.cs]: I don't think we should provide this overload.

File: src/Microsoft.AspNetCore.Razor.Language/RazorLanguageVersion.cs [hunk 6]
```diff
@@ -1,33 +1,33 @@
...
 
+using System;
+
 namespace Microsoft.AspNetCore.Razor.Language
 {
-    public enum RazorLanguageVersion
+    public sealed class RazorLanguageVersion
     {
-        Version1_0 = 1,
+        public static readonly RazorLanguageVersion Version1_0 = new RazorLanguageVersion(1, 0);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorLanguageVersion.cs]: Suggestion: name these like `Version_1_0` instead. That reads a little nicer and it's what I did in MVC for this kind of thing.

File: src/Microsoft.AspNetCore.Razor.Language/RazorLanguageVersion.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Razor.Language/RazorLanguageVersion.cs]: Yeah, this will eventually have a `TryParse` method.

File: src/Microsoft.AspNetCore.Razor.Language/RazorParserOptions.cs [hunk 7]
```diff
@@ -8,15 +8,13 @@ namespace Microsoft.AspNetCore.Razor.Language
 {
     public abstract class RazorParserOptions
     {
-        internal static readonly RazorLanguageVersion LatestRazorLanguageVersion = RazorLanguageVersion.Version2_1;
-
         public static RazorParserOptions CreateDefault()
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorParserOptions.cs]: We might end up obsoleting some of these things when the day is through.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs [hunk 8]
```diff
@@ -51,10 +58,12 @@ public override bool Equals(ProjectExtensibilityConfiguration other)
...
-            return Enumerable.SequenceEqual(
-                Assemblies.OrderBy(a => a.Identity.Name).Select(a => a.Identity),
-                other.Assemblies.OrderBy(a => a.Identity.Name).Select(a => a.Identity),
-                AssemblyIdentityEqualityComparer.NameAndVersion);
+            return 
+                Enumerable.SequenceEqual(
+                    Assemblies.OrderBy(a => a.Identity.Name).Select(a => a.Identity),
+                    other.Assemblies.OrderBy(a => a.Identity.Name).Select(a => a.Identity),
+                    AssemblyIdentityEqualityComparer.NameAndVersion) &&
+                RazorLanguageVersion == other.RazorLanguageVersion;
```
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs]: Was this just a bug?

Comment (NTaylorMullen) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs]: No, just formatted and added the `RazorLanguageVersion` check to the equality.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs [hunk 9]
```diff
@@ -28,19 +30,24 @@ internal class MvcExtensibilityConfiguration : ProjectExtensibilityConfiguration
...
+        // MVC: '2.0.0' (fallback) | Razor Language '2.0.0'
+        // or 
+        // MVC: '2.1.3' | Razor Language '2.1.3'
+        public override string DisplayName => $"MVC: {MvcAssembly.Identity.Version.ToString(3)}" + (Kind == ProjectExtensibilityConfigurationKind.Fallback? " (fallback)" : string.Empty) + " | " + RazorLanguageVersion;
 
         public override ProjectExtensibilityConfigurationKind Kind { get; }
 
         public override ProjectExtensibilityAssembly RazorAssembly { get; }
 
+        public override RazorLanguageVersion RazorLanguageVersion { get; }
```
Comment (NTaylorMullen) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs]: Exposing `RazorLanguageVersion` here and not `RazorConfiguration` because I'm not convinced that we should tie projects to runtime/design time. Someone else should make that decision depending on how this extensibility configuration is consumed.

**Comment by NTaylorMullen** [comment]
🆙 📅

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs (hunk 9 above)
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs]: I'm not sure I understand the rationale behind that, but this seems like a safe minimal change for now. We'll need to revisit this class again in the near future, so we'll eventually do the right thing anyway 😆. Let's discuss when we get back to this spot.

File: src/Microsoft.AspNetCore.Razor.Language/RazorLanguageVersion.cs [hunk 10]
```diff
@@ -1,33 +1,33 @@
...
-            return false;
+            Major = major;
+            Minor = minor;
         }
+
+        public int Major { get; }
+
+        public int Minor { get; }
+
+        public override string ToString() => $"Razor Language '{Major}.{Minor}'";
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorLanguageVersion.cs]: Suggest taking `Language` out of this. Including `Language` makes it seem like it should be localized.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs [hunk 11]
```diff
@@ -4,13 +4,15 @@
...
 using System.Linq;
+using Microsoft.AspNetCore.Razor.Language;
 using Microsoft.Extensions.Internal;
 
 namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
 {
     internal class MvcExtensibilityConfiguration : ProjectExtensibilityConfiguration
     {
         public MvcExtensibilityConfiguration(
+            RazorLanguageVersion razorLanguageVersion,
```
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs]: languageVersion

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs (hunk 8 above)
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs]: Ok gotcha

Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs]: nit: can you put the language version first? That should be very cheap to compare.

Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/MvcExtensibilityConfiguration.cs]: another non-nit. Either make `RazorLanguageVersion` implement `IEquatable<>` or use object.ReferenceEquals(...). Using `==` with non-simple types is non-obvious what it does.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectExtensibilityConfiguration.cs [hunk 12]
```diff
@@ -16,6 +17,8 @@ internal abstract class ProjectExtensibilityConfiguration : IEquatable<ProjectEx
 
         public abstract ProjectExtensibilityAssembly RazorAssembly { get; }
 
+        public abstract RazorLanguageVersion RazorLanguageVersion { get; }
```
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectExtensibilityConfiguration.cs]: LanguageVersion

File: src/Microsoft.VisualStudio.Editor.Razor/DefaultTemplateEngineFactoryService.cs [hunk 13]
```diff
@@ -41,11 +42,13 @@ public override RazorTemplateEngine Create(string projectPath, Action<IRazorEngi
...
             var configuration = (project?.Configuration as MvcExtensibilityConfiguration) ?? DefaultConfiguration;
+            var razorLanguageVersion = configuration.RazorLanguageVersion;
+            var razorConfiguration = new RazorConfiguration(razorLanguageVersion, designTime: true);
 
             RazorEngine engine;
-            if (configuration.RazorAssembly.Identity.Version.Major == 1)
+            if (razorLanguageVersion.Major == 1)
             {
-                engine = RazorEngine.CreateDesignTime(b =>
+                engine = RazorEngine.CreateCore(razorConfiguration, b =>
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.Editor.Razor/DefaultTemplateEngineFactoryService.cs]: Looks good. I assume this is the meat of it. The pieces come together pretty well.
